sbnanalysis/ana/SBNOsc/NueSelection.cxx

```diff
@@ -2,60 +2,345 @@
+    for (auto j : EnergeticShowersIndices) {
+      auto const& shower = mcshowers.at(j);
+      auto shower_pos = shower.DetProfile().Position();
+      double distance = (nu_pos.Vect()-shower_pos.Vect()).Mag();
+      fDiffLength->Fill(distance);
+      if (distance <= 5.) {
+        matched_shower_count++;
+        assn_showers_pdg.push_back(shower.PdgCode());
+        assn_showers_quality.push_back(HasGooddEdx[j]);
+      }
+    }
+    if (!assn_showers_pdg.empty()) ShowerPDG.push_back(assn_showers_pdg[0]);
+    else ShowerPDG.push_back(0);
+    if (!assn_showers_quality.empty()) AssnShowerGooddEdx.push_back(assn_showers_quality[0]);
+    else AssnShowerGooddEdx.push_back(false);
+    if (matched_shower_count>0) matchedness.push_back(true);
+    else matchedness.push_back(false);
+  }
+  assert(ShowerPDG.size()==matchedness.size());
+  assert(AssnShowerGooddEdx.size()==matchedness.size());
+
+  // Iterate through the neutrinos/MCTruth
   for (size_t i=0; i<mctruths.size(); i++) {
     Event::Interaction interaction;
     auto const& mctruth = mctruths.at(i);
     const simb::MCNeutrino& nu = mctruth.GetNeutrino();
+    auto nu_E = nu.Nu().E();
+    auto vx = nu.Nu().Vx();
+    auto vy = nu.Nu().Vy();
+    auto vz = nu.Nu().Vz();
+    bool IsFid = (((-174.15 < vx && vx < -27.65) || (27.65 < vx && vx < 174.15)) && (-175 < vy && vy < 175) && (25 < vz && vz < 475));
+    int MuTrackCount=0;
+    auto nu_pos = nu.Nu().Position();
+    for (size_t j=0;j<mctracks.size();j++) {
+      auto const& mctrack = mctracks.at(j);
+      auto track_start_pos = mctrack.Start().Position();
+      auto nu_track_dis = (track_start_pos.Vect() - nu_pos.Vect()).Mag();
+      //do total length cut first
+      auto total_length = (mctrack.Start().Position().Vect() - mctrack.End().Position().Vect()).Mag();
+      if ((total_length >= 100.) && (nu_track_dis<=5.)) MuTrackCount++;
+    }
+    bool NotMuTrackness = (MuTrackCount==0);
+
+    if (matchedness[i]&&IsFid) fShowerCutSelectionHist->Fill(nu_E);
+    if (matchedness[i]&&IsFid&&NotMuTrackness) fSelectedNuHist->Fill(nu_E);
+    if (matchedness[i]&&IsFid&&NotMuTrackness&&PassConversionGap[i]) fNodEdxNuHist->Fill(nu_E);
+    if (matchedness[i]&&IsFid&&NotMuTrackness&&PassConversionGap[i]&&AssnShowerGooddEdx[i]) {
+      fCGSelectionHist->Fill(nu_E);
+      if (nu.Nu().PdgCode() ==12) fSelectedTrueNue->Fill(nu_E);
+
+      //fill in shower pdg histograms
+      if (ShowerPDG[i]==13) fMuShowerSelectedNu->Fill(nu_E);
+      if (ShowerPDG[i]==11) fEShowerSelectedNu->Fill(nu_E);
+      if (ShowerPDG[i]==22) fGammaShowerSelectedNu->Fill(nu_E);
+      if ((ShowerPDG[i]!=13)&&(ShowerPDG[i]!=11)&&(ShowerPDG[i]!=22))
+        fOtherShowerSelectedNu->Fill(nu_E);
+
+      double lower_bound=0.;
+      double upper_bound=1.;
+      std::uniform_real_distribution<double> unif(lower_bound,upper_bound);
+      std::random_device r;
+      std::default_random_engine e1(r());
+      double rnd_double = unif(e1);
```

Comment:
Would it be possible to handle this by including a weight < 1 for the event somewhere, rather than rejecting it in the selection? That should achieve the same effect in the end, but preserve as much of the MC statistics as possible.

(And in general for random numbers, I'd prefer to stick with the ROOT random number generator (`#include <TRandom.h>` and use `gRandom`, which is a `TRandom3` object) for consistency.)